### PR TITLE
Fix OIDC logout

### PR DESCRIPTION
**Pull request by julienhuon**
Hello Misagh,

The OIDC RP-initiated logout [describe here](https://openid.net/specs/openid-connect-session-1_0.html#RPLogout) seems broken in CAS if you're using any of the specified parameters.

This pull request fixes the oidc logout and adds unit tests to avoid any future problem. the following bugs are fixed:

### State Parameter
The state parameter is never returned:

```
> GET /cas/oidc/logout?state=fefefe HTTP/1.1
< Location: http://cas.example.org:8080/cas/logout?state
```

### RedirectParameter is not defaulted to service
If you're not setting the following property in your cas.properties:

```
cas.logout.redirectParameter=service
```

When you're sending a valid id token, the following error will appears:

```
2020-02-21 11:53:16,056 DEBUG [org.apereo.cas.logout.slo.DefaultSingleLogoutServiceLogoutUrlBuilder] - <Logout request will be sent to [http://logout,http://pouet] for service [AbstractWebApplicationService(id=svod, originalUrl=svod, artifactId=null, principal=null, source=null, loggedOutAlready=false, format=XML, attributes={})]>
2020-02-21 11:53:16,071 DEBUG [org.apereo.cas.web.FlowExecutionExceptionResolver] - <Ignoring the received exception [java.lang.IllegalArgumentException: Name must not be null] due to a type mismatch with handler [org.apereo.cas.oidc.web.controllers.logout.OidcLogoutEndpointController#handleRequestInternal(String, String, String, HttpServletRequest, HttpServletResponse)]>
2020-02-21 11:53:16,073 DEBUG [org.apereo.cas.web.FlowExecutionExceptionResolver] - <Ignoring the received exception [java.lang.IllegalArgumentException: Name must not be null] due to a type mismatch with handler [org.apereo.cas.oidc.web.controllers.logout.OidcLogoutEndpointController#handleRequestInternal(String, String, String, HttpServletRequest, HttpServletResponse)]>
2020-02-21 11:53:16,076 ERROR [org.springframework.boot.web.servlet.support.ErrorPageFilter] - <Forwarding to error page from request [/oidc/logout] due to exception [Name must not be null]>

```

And you will never be redirected.

### The redirect url is not correctly formatted when you're not using post_logout_redirect_uri

If you're sending a request with an id_token but without post_logout_redirect_uri, cas will redirect you to an invalid url:

```
> GET /cas/oidc/logout?id_token_hint=eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6InBvdWV0In0.eyJqdGkiOiJUR1QtMS1rZFZmOGMyU1BsZmJCMFFURU9LNllaNWMxVVpKNDJIUGpFbW10UnluSTFNMFBkaXpaa2hLYXFYaWNmay1HNjdOLW9nLWEwZmYwYTAxYzVhYiIsImlzcyI6Imh0dHA6Ly9jYXMuZXhhbXBsZS5vcmc6ODA4MC9jYXMvb2lkYyIsImF1ZCI6InN2b2QiLCJleHAiOjE1ODIzMTYxNTMsImlhdCI6MTU4MjI4NzM1MywibmJmIjoxNTgyMjg3MDUzLCJzdWIiOiJwb3VldCIsImNsaWVudF9pZCI6InN2b2QiLCJhdXRoX3RpbWUiOjE1ODIyODczNDUsInN0YXRlIjoiRTRIM0NJOUM5elQ5T0ZBeDd5dU9FZyIsIm5vbmNlIjoiIiwiYXRfaGFzaCI6Ii01ZXF6V1JLWTZYZ0g5Zm9PQTFXWUEiLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJwb3VldCJ9.1_0Kdo6h_Pw73Z5aN0OOnt1sVLXPX1AMwy4Er2o1ZB-wtX6Ua3uOSFsFsSQKUMf6jRsi0JwfdB44ICiVvKfVQQ&state=fefefe HTTP/1.1

< Location: http://cas.example.org:8080/cas/logout?service=org.apereo.cas.logout.slo.SingleLogoutUrl@b123286f&state=org.apereo.cas.logout.slo.SingleLogoutUrl@b123286f
```

Notice the state parameter value.

Regards,
Julien

### CODE REVIEW

File: api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/logout/LogoutProperties.java [hunk 1]
```diff
@@ -25,7 +26,7 @@
      * is indicated and extracted by a parameter name of your choosing here. If none specified,
      * the default will be used as {@code service}.
      */
-    private String redirectParameter;
+    private String redirectParameter = CasProtocolConstants.PARAMETER_SERVICE;
```
Comment (mmoayyed) [inline api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/logout/LogoutProperties.java]: Please revert this change; it's preferable to not depend on `CasProtocolConstants` if possible.

File: support/cas-server-support-actions/src/main/java/org/apereo/cas/web/flow/logout/LogoutAction.java [hunk 2]
```diff
@@ -52,7 +51,7 @@ protected Event doInternalExecute(final HttpServletRequest request, final HttpSe
             () -> Boolean.FALSE)
             .get();
 
-        val paramName = StringUtils.defaultIfEmpty(logoutProperties.getRedirectParameter(), CasProtocolConstants.PARAMETER_SERVICE);
```
Comment (mmoayyed) [inline support/cas-server-support-actions/src/main/java/org/apereo/cas/web/flow/logout/LogoutAction.java]: Please revert this; See above.

File: api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/logout/LogoutProperties.java (hunk 1 above)
Comment (julienhuon) [reply api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/logout/LogoutProperties.java]: Hi,

So, should I just default it like this ?

```
private String redirectParameter = "service"
```

If it is not defined, the second bug described in this PR will occurs. Or should I default it in the OidcLogoutEndpointController.java file ?

Regards,
Julien

Comment (mmoayyed) [reply api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/logout/LogoutProperties.java]: The second option, where you default it much like the LogoutAction, as:

```
StringUtils.defaultIfEmpty(logoutProperties.getRedirectParameter(), CasProtocolConstants.PARAMETER_SERVICE);
```

Comment (mmoayyed) [reply api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/logout/LogoutProperties.java]: On second thought using `"service"` as the default is perfectly fine too :)

Comment (julienhuon) [reply api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/logout/LogoutProperties.java]: Ok.

Will do it in a few hours. :)

Comment (julienhuon) [reply api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/core/logout/LogoutProperties.java]: Done.
